shared/src/Datadog.Trace.ClrProfiler.Native/crashhandler.cpp

```diff
@@ -0,0 +1,409 @@
+
+        return result;
+    }
+
+    std::unique_ptr<CrashHandler> CrashHandler::Create()
+    {
+        std::unique_ptr<CrashHandler> crashHandler(new CrashHandler());
+
+        if (crashHandler->Register())
+        {
+            return crashHandler;
+        }
+
+        return nullptr;
+    }
+
+    CrashHandler::~CrashHandler()
+    {
+        if (!_crashHandler.empty())
+        {
+            auto hr = WerUnregisterRuntimeExceptionModule(_crashHandler.c_str(), &_context);
+            Log::Debug("Crashtracking - Unregistering crash handler: ", hr);
+            _crashHandler.clear();
+        }
+
+        if (_context.Environ != nullptr)
+        {
+            delete[] _context.Environ;
+        }
+
+        _context.Environ = nullptr;
+        _context.EnvironLength = 0;
+    }
+
+    CrashHandler::CrashHandler()
+        : _context(),
+        _crashHandler()
+    {
+    }
+
+    bool CrashHandler::Register()
+    {
+        if (!_crashHandler.empty())
+        {
+            Log::Warn("Crashtracking - Trying to initialize the crash handler twice");
+            return false;
+        }
+
+        auto dllPath = GetCurrentDllPath();
+
+        if (dllPath.empty())
+        {
+            Log::Warn("Crashtracking - Could not register the crash handler: error when retrieving the path of the DLL");
+            return false;
+        }
+
+        GetDatadogEnvironmentBlock(_context.Environ, _context.EnvironLength);
+
+        // Register the crash handler in the registry
+        // Windows expects a DWORD value with the full path of the DLL as the name, 
+        // in SOFTWARE\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules.
+        // The key can be located either in HKLM or HKCU. The MSI will create the value in HKLM,
+        // but if it's missing we add it to HKCU.
+        HKEY hKey;
+        LPCWSTR subKey = L"SOFTWARE\\Microsoft\\Windows\\Windows Error Reporting\\RuntimeExceptionHelperModules";
+        DWORD value = 1;
+
+        if (!RegistryValueExists(HKEY_LOCAL_MACHINE, subKey, dllPath.c_str()))
+        {
+            // Open the key
+            DWORD disposition;
+            auto result = RegCreateKeyEx(HKEY_CURRENT_USER, subKey, 0, NULL, 0, KEY_SET_VALUE, NULL, &hKey, &disposition);
+
+            if (result != ERROR_SUCCESS)
+            {
+                // Failing to set the registry is not a fatal error: in the worst case scenario the crash handler just won't be invoked by WER
+                Log::Warn("Crashtracking - Failed to create registry key: ", GetLastError());
+            }
+            else
+            {
+                // Set the value
+                result = RegSetValueEx(hKey, dllPath.c_str(), 0, REG_DWORD, reinterpret_cast<const BYTE*>(&value), sizeof(value));
+
+                if (result != ERROR_SUCCESS)
+                {
+                    Log::Warn("Crashtracking - Failed to set registry value: ", GetLastError());
```

Comment:
nit: is it worth using a slightly different error message here to distinguish between the above error?

Reply:
They're already different? (failed to create/failed to set)